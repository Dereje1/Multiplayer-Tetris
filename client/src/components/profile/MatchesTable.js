import * as React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
/* MUI */
import Box from '@mui/material/Box';
import Collapse from '@mui/material/Collapse';
import IconButton from '@mui/material/IconButton';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Typography from '@mui/material/Typography';
import Paper from '@mui/material/Paper';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';
import EmojiEventsIcon from '@mui/icons-material/EmojiEvents';
import SentimentVeryDissatisfiedIcon from '@mui/icons-material/SentimentVeryDissatisfied';

const getStatusIcon = (isWinner) => (
  isWinner ? <EmojiEventsIcon sx={{ color: 'green', fontSize: 26 }} />
    :
    <SentimentVeryDissatisfiedIcon sx={{ color: 'red', fontSize: 26 }} />
)

const extractRowInfo = ({ row, opponents, userId }) => {
  const isWinner = userId === row.winnerGoogleId
  const opponentId = isWinner ? row.looserGoogleId : row.winnerGoogleId
  const details = ['You', opponents[opponentId]].map((player, idx) => {
    if (idx === 0) {
      return isWinner ?
        {
          player,
          linesCleared: row.winnerLinesCleared,
          floorsRaised: row.winnerFloorsRaised,
          isDisqualified: 'No'
        }
        :
        {
          player,
          linesCleared: row.looserLinesCleared,
          floorsRaised: row.looserFloorsRaised,
          isDisqualified: row.looserDisqualified ? 'Yes' : 'No'
        }
    }

    return isWinner ?
      {
        player,
        linesCleared: row.looserLinesCleared,
        floorsRaised: row.looserFloorsRaised,
        isDisqualified: row.looserDisqualified ? 'Yes' : 'No'
      }
      :
      {
        player,
        linesCleared: row.winnerLinesCleared,
        floorsRaised: row.winnerFloorsRaised,
        isDisqualified: 'No'
      }
  })
  return {
    ...row,
    opponentName: opponents[opponentId],
    isWinner: getStatusIcon(isWinner),
    details
  }
}

export const Row = ({ row, opponents, userId }) => {
  const [open, setOpen] = React.useState(false);
  const rowInfo = extractRowInfo({ row, opponents, userId })
  return (
    <React.Fragment>
      <TableRow sx={{ '& > *': { borderBottom: 'unset' } }}>
        <TableCell>
          <IconButton
            aria-label="expand row"
            size="small"
            onClick={() => setOpen(!open)}
          >
            {open ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
          </IconButton>
        </TableCell>
        <TableCell component="th" scope="row">
          {moment.utc(rowInfo.createdAt).format('MMMM Do YYYY, h:mm a')}
        </TableCell>
        <TableCell align="center">{rowInfo.opponentName}</TableCell>
        <TableCell align="center">{rowInfo.isWinner}</TableCell>
        <TableCell align="center">{rowInfo.difficulty}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={6}>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <Box sx={{ margin: 1 }}>
              <Typography  gutterBottom component="div" sx={{fontWeight:'bold'}}>
                Details
              </Typography>
              <Table size="small" aria-label="purchases">
                <TableHead>
                  <TableRow>
                    <TableCell></TableCell>
                    <TableCell align="center">Lines cleared</TableCell>
                    <TableCell align="center">Floors raised</TableCell>
                    <TableCell align="center">Disqualified</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {rowInfo.details.map((detailsRow, idx) => (
                    <TableRow key={`${rowInfo._id}_${idx}`}>
                      <TableCell component="th" scope="row">
                        {detailsRow.player}
                      </TableCell>
                      <TableCell align="center">{detailsRow.linesCleared}</TableCell>
                      <TableCell align="center">{detailsRow.floorsRaised}</TableCell>
                      <TableCell align="center">
                        {detailsRow.isDisqualified}
                      </TableCell>
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            </Box>
          </Collapse>
        </TableCell>
      </TableRow>
    </React.Fragment>
  );
}

export default function MatchesTable({ rows, opponents, userId }) {
  return (
    <TableContainer component={Paper}>
      <Table aria-label="collapsible table">
        <TableHead>
          <TableRow>
            <TableCell />
            <TableCell>Date (UTC)</TableCell>
            <TableCell align="center">Opponent</TableCell>
            <TableCell align="center">Win/Loss</TableCell>
            <TableCell align="center">Difficulty level</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row) => (
            <Row
              key={row._id}
              row={row}
              opponents={opponents}
              userId={userId}
            />
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
}

MatchesTable.propTypes = {
  rows:PropTypes.arrayOf(PropTypes.object).isRequired,
  opponents: PropTypes.object.isRequired,
  userId:PropTypes.string.isRequired
}

Row.propTypes = {
  row: PropTypes.object.isRequired,
  opponents: PropTypes.object.isRequired,
  userId:PropTypes.string.isRequired
}