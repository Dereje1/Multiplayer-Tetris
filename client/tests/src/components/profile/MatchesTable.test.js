import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import MatchesTable, { Row } from '../../../../src/components/profile/MatchesTable'
import { stubProfile } from '../../../stub'

//had to mock usestate as jest could not support due to conflicting react
//TODO: fix
const mockSetOpen = jest.fn()
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    useState: jest.fn(() => [false, mockSetOpen])
}))
// eslint-disable-next-line import/first
import { useState } from 'react';

describe('The Matches table', () => {
    test('will render', () => {
        const wrapper = shallow(<MatchesTable
            rows={stubProfile.userProfileResponse.matchStats}
            opponents={stubProfile.userProfileResponse.opponentNames}
            userId={stubProfile.user.profile.username}
        />)
        expect(toJson(wrapper)).toMatchSnapshot();
    })
})

describe('The Matches table row', () => {
    test('will render a collapsed row for a winning user', () => {
        const props = {
            row: stubProfile.userProfileResponse.matchStats[0],
            opponents: stubProfile.userProfileResponse.opponentNames,
            userId: stubProfile.user.profile.username
        }
        const wrapper = shallow(<Row {...props} />)
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will render a collapsed row for a loosing user', () => {
        const props = {
            row: stubProfile.userProfileResponse.matchStats[1],
            opponents: stubProfile.userProfileResponse.opponentNames,
            userId: stubProfile.user.profile.username
        }
        const wrapper = shallow(<Row {...props} />)
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will render an expanded row', () => {
        useState.mockImplementation(()=>[true, mockSetOpen])
        const props = {
            row: stubProfile.userProfileResponse.matchStats[0],
            opponents: stubProfile.userProfileResponse.opponentNames,
            userId: stubProfile.user.profile.username
        }
        const wrapper = shallow(<Row {...props} />)
        const expandButton = wrapper.find({ 'aria-label': "expand row" })
        expandButton.props().onClick()
        expect(toJson(wrapper)).toMatchSnapshot();
    })

    test('will change state of an expanded row', () => {
        useState.mockImplementation(()=>[false, mockSetOpen])
        const props = {
            row: stubProfile.userProfileResponse.matchStats[0],
            opponents: stubProfile.userProfileResponse.opponentNames,
            userId: stubProfile.user.profile.username
        }
        const wrapper = shallow(<Row {...props} />)
        const expandButton = wrapper.find({ 'aria-label': "expand row" })
        expandButton.props().onClick()
        expect(mockSetOpen).toHaveBeenCalledWith(true)
    })
})