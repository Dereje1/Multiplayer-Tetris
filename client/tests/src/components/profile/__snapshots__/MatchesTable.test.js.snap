// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Matches table row will render a collapsed row for a loosing user 1`] = `
<Fragment>
  <ForwardRef(TableRow)
    sx={
      Object {
        "& > *": Object {
          "borderBottom": "unset",
        },
      }
    }
  >
    <ForwardRef(TableCell)>
      <ForwardRef(IconButton)
        aria-label="expand row"
        onClick={[Function]}
        size="small"
      >
        <Memo(ForwardRef(KeyboardArrowDownIcon)) />
      </ForwardRef(IconButton)>
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      component="th"
      scope="row"
    >
      July 23rd 2022, 9:08 pm
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      Stub oponent name
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      <Memo(ForwardRef(SentimentVeryDissatisfiedIcon))
        sx={
          Object {
            "color": "red",
            "fontSize": 26,
          }
        }
      />
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      4
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
  <ForwardRef(TableRow)>
    <ForwardRef(TableCell)
      colSpan={6}
      style={
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <ForwardRef(Collapse)
        in={false}
        timeout="auto"
        unmountOnExit={true}
      >
        <ForwardRef(Box)
          sx={
            Object {
              "margin": 1,
            }
          }
        >
          <ForwardRef(Typography)
            component="div"
            gutterBottom={true}
            sx={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Details
          </ForwardRef(Typography)>
          <ForwardRef(Table)
            aria-label="purchases"
            size="small"
          >
            <ForwardRef(TableHead)>
              <ForwardRef(TableRow)>
                <ForwardRef(TableCell) />
                <ForwardRef(TableCell)
                  align="center"
                >
                  Lines cleared
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Floors raised
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Disqualified
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableHead)>
            <ForwardRef(TableBody)>
              <ForwardRef(TableRow)
                key="62dc633b2464e19567160cb3_0"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  You
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  No
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
              <ForwardRef(TableRow)
                key="62dc633b2464e19567160cb3_1"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  Stub oponent name
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  2
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  1
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  No
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableBody)>
          </ForwardRef(Table)>
        </ForwardRef(Box)>
      </ForwardRef(Collapse)>
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
</Fragment>
`;

exports[`The Matches table row will render a collapsed row for a winning user 1`] = `
<Fragment>
  <ForwardRef(TableRow)
    sx={
      Object {
        "& > *": Object {
          "borderBottom": "unset",
        },
      }
    }
  >
    <ForwardRef(TableCell)>
      <ForwardRef(IconButton)
        aria-label="expand row"
        onClick={[Function]}
        size="small"
      >
        <Memo(ForwardRef(KeyboardArrowDownIcon)) />
      </ForwardRef(IconButton)>
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      component="th"
      scope="row"
    >
      July 24th 2022, 8:37 am
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      Stub oponent name
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      <Memo(ForwardRef(EmojiEventsIcon))
        sx={
          Object {
            "color": "green",
            "fontSize": 26,
          }
        }
      />
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      2
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
  <ForwardRef(TableRow)>
    <ForwardRef(TableCell)
      colSpan={6}
      style={
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <ForwardRef(Collapse)
        in={false}
        timeout="auto"
        unmountOnExit={true}
      >
        <ForwardRef(Box)
          sx={
            Object {
              "margin": 1,
            }
          }
        >
          <ForwardRef(Typography)
            component="div"
            gutterBottom={true}
            sx={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Details
          </ForwardRef(Typography)>
          <ForwardRef(Table)
            aria-label="purchases"
            size="small"
          >
            <ForwardRef(TableHead)>
              <ForwardRef(TableRow)>
                <ForwardRef(TableCell) />
                <ForwardRef(TableCell)
                  align="center"
                >
                  Lines cleared
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Floors raised
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Disqualified
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableHead)>
            <ForwardRef(TableBody)>
              <ForwardRef(TableRow)
                key="62dd04cb1a1a257ead39a1ea_0"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  You
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  No
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
              <ForwardRef(TableRow)
                key="62dd04cb1a1a257ead39a1ea_1"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  Stub oponent name
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Yes
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableBody)>
          </ForwardRef(Table)>
        </ForwardRef(Box)>
      </ForwardRef(Collapse)>
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
</Fragment>
`;

exports[`The Matches table row will render an expanded row 1`] = `
<Fragment>
  <ForwardRef(TableRow)
    sx={
      Object {
        "& > *": Object {
          "borderBottom": "unset",
        },
      }
    }
  >
    <ForwardRef(TableCell)>
      <ForwardRef(IconButton)
        aria-label="expand row"
        onClick={[Function]}
        size="small"
      >
        <Memo(ForwardRef(KeyboardArrowUpIcon)) />
      </ForwardRef(IconButton)>
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      component="th"
      scope="row"
    >
      July 24th 2022, 8:37 am
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      Stub oponent name
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      <Memo(ForwardRef(EmojiEventsIcon))
        sx={
          Object {
            "color": "green",
            "fontSize": 26,
          }
        }
      />
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      align="center"
    >
      2
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
  <ForwardRef(TableRow)>
    <ForwardRef(TableCell)
      colSpan={6}
      style={
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <ForwardRef(Collapse)
        in={true}
        timeout="auto"
        unmountOnExit={true}
      >
        <ForwardRef(Box)
          sx={
            Object {
              "margin": 1,
            }
          }
        >
          <ForwardRef(Typography)
            component="div"
            gutterBottom={true}
            sx={
              Object {
                "fontWeight": "bold",
              }
            }
          >
            Details
          </ForwardRef(Typography)>
          <ForwardRef(Table)
            aria-label="purchases"
            size="small"
          >
            <ForwardRef(TableHead)>
              <ForwardRef(TableRow)>
                <ForwardRef(TableCell) />
                <ForwardRef(TableCell)
                  align="center"
                >
                  Lines cleared
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Floors raised
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Disqualified
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableHead)>
            <ForwardRef(TableBody)>
              <ForwardRef(TableRow)
                key="62dd04cb1a1a257ead39a1ea_0"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  You
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  No
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
              <ForwardRef(TableRow)
                key="62dd04cb1a1a257ead39a1ea_1"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  Stub oponent name
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  0
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)
                  align="center"
                >
                  Yes
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableBody)>
          </ForwardRef(Table)>
        </ForwardRef(Box)>
      </ForwardRef(Collapse)>
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
</Fragment>
`;

exports[`The Matches table will render 1`] = `
<ForwardRef(TableContainer)
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "children": [Function],
        "className": [Function],
        "classes": [Function],
        "component": [Function],
        "elevation": [Function],
        "square": [Function],
        "sx": [Function],
        "variant": [Function],
      },
      "render": [Function],
    }
  }
>
  <ForwardRef(Table)
    aria-label="collapsible table"
  >
    <ForwardRef(TableHead)>
      <ForwardRef(TableRow)>
        <ForwardRef(TableCell) />
        <ForwardRef(TableCell)>
          Date (UTC)
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)
          align="center"
        >
          Opponent
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)
          align="center"
        >
          Win/Loss
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)
          align="center"
        >
          Difficulty level
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
    </ForwardRef(TableHead)>
    <ForwardRef(TableBody)>
      <Row
        key="62dd04cb1a1a257ead39a1ea"
        opponents={
          Object {
            "0002": "Stub oponent name",
          }
        }
        row={
          Object {
            "__v": 0,
            "_id": "62dd04cb1a1a257ead39a1ea",
            "createdAt": "2022-07-24T08:37:31.809Z",
            "difficulty": 2,
            "looserDisqualified": true,
            "looserFloorsRaised": 0,
            "looserGoogleId": "0002",
            "looserLinesCleared": 0,
            "updatedAt": "2022-07-24T08:37:31.809Z",
            "winnerFloorsRaised": 0,
            "winnerGoogleId": "0001",
            "winnerLinesCleared": 0,
          }
        }
        userId="0001"
      />
      <Row
        key="62dc633b2464e19567160cb3"
        opponents={
          Object {
            "0002": "Stub oponent name",
          }
        }
        row={
          Object {
            "__v": 0,
            "_id": "62dc633b2464e19567160cb3",
            "createdAt": "2022-07-23T21:08:11.812Z",
            "difficulty": 4,
            "looserDisqualified": false,
            "looserFloorsRaised": 0,
            "looserGoogleId": "0001",
            "looserLinesCleared": 0,
            "updatedAt": "2022-07-23T21:08:11.812Z",
            "winnerFloorsRaised": 1,
            "winnerGoogleId": "0002",
            "winnerLinesCleared": 2,
          }
        }
        userId="0001"
      />
    </ForwardRef(TableBody)>
  </ForwardRef(Table)>
</ForwardRef(TableContainer)>
`;
